File: examples/monorepo_example/elixir/lib/monorepo_example/jobs/cross_language_concurrency_job.ex
defmodule MonorepoExample.Jobs.CrossLanguageConcurrencyJob do
  use GoodJob.Job, queue: "ex.default"

  @impl GoodJob.Behaviour
  def perform(%{resource_id: resource_id, action: action}) do
    require Logger
    Logger.info("CrossLanguageConcurrencyJob: resource_id=#{resource_id}, action=#{action}")
    Process.sleep(2000)

    Logger.info("CrossLanguageConcurrencyJob: Completed resource_id=#{resource_id}")
    :ok
  end

  def perform(args) when is_map(args) do
    resource_id = Map.get(args, :resource_id) || Map.get(args, "resource_id") || "unknown"
    action = Map.get(args, :action) || Map.get(args, "action") || "process"
    perform(%{resource_id: resource_id, action: action})
  end

  def good_job_concurrency_config do
    [total_limit: 2]
  end
end

File: examples/monorepo_example/rails/app/jobs/cross_language_concurrency_job.rb
class CrossLanguageConcurrencyJob < ApplicationJob
  include GoodJob::ActiveJobExtensions::Concurrency

  # This job can be processed by either Ruby or Elixir workers
  # The concurrency limit applies across BOTH languages

  # Queue can be processed by either worker
  queue_as "ex.default"  # Processed by Elixir worker
  # Or use "rb.default" for Ruby worker

  # Concurrency configuration - limits apply across all workers
  good_job_control_concurrency_with(
    total_limit: 2,  # Maximum 2 concurrent jobs across ALL workers (Ruby + Elixir)
    key: -> {
      # Generate concurrency key from job arguments
      args = arguments.first || {}
      key = args[:resource_id] || args["resource_id"] || "default"
      "resource:#{key}"
    }
  )

  def perform(resource_id:, action: "process")
    # This will be processed by Elixir worker
    # But concurrency limits are enforced across Ruby and Elixir
    Rails.logger.info "CrossLanguageConcurrencyJob: resource_id=#{resource_id}, action=#{action}"
    puts "[Cross-Language] Processing resource #{resource_id} with action #{action}"

    # Simulate work
    sleep(2)

    Rails.logger.info "CrossLanguageConcurrencyJob: Completed resource_id=#{resource_id}"
  end
end

